### PR TITLE
Add student account settings page

### PR DESCRIPTION
**Pull request by guyy13**
Add the option for student users to edit their account settings.
Some more details:
- Each student can only access their own account settings page once they are not logged in, trying to get to the edit student account settings link without being logged in as a student will send you to "dead_end.html"
- Edited the base.html to contain a link to the edit account setting page of a Student (only visible if a student is logged in)
- Added a success page

### CODE REVIEW

**Review by YarinBou — APPROVED** [review]
lgtm

**Review by yaelerdr — APPROVED** [review]
good job!! looks good

**Review by eldaraharonson — APPROVED** [review]

**Review by benmenachem — APPROVED** [review]

File: student/views.py
Comment (guy9050) [inline student/views.py]: what do you test here?

File: student/views.py
Comment (guy9050) [inline student/views.py]: what do you need this for?

File: student/views.py
Comment (guy9050) [inline student/views.py]: why are you setting it?

File: student/views.py
Comment (guyy13) [reply student/views.py]: @guy9050 
It's not a test (like a pytest test).
The abstract class 'UserPassesTestMixin' allows you to render a view only if the user that sent the request passes a certain test. the 'test_func' method is an abstract method of this class that has to be implemented when inheriting from it.
It checks here that the ID of the user that sent the request is equal to the ID in the URL that they're trying to get to.
This way you can only edit the account details of your own account, and you cannot access any other accounts by modifying the URL.

File: student/views.py
Comment (guyy13) [reply student/views.py]: @guy9050 
I needed to check within the HTML whether the logged in user is a Student or a Recruiter.
I couldn't find a way to call a python function from within the HTML, so the solution I found was to pass the user role in the context dictionary to the HTML page.
In this function, I get the context dictionary that is being sent to the page and add an 'is_student' key, this way I can check from within the HTML whether the logged in user is a Student, and display the correct links in the navigation bar.

File: student/views.py
Comment (guyy13) [reply student/views.py]: See my comment above

**Review by YarinBou — APPROVED** [review]

File: student/views.py
Comment (guy9050) [reply student/views.py]: thanks @guyy13

File: student/views.py
Comment (guy9050) [reply student/views.py]: I remember there were functions like `is_recruiter()` or `is_student()`, can't you use those? what do you mean by needed to check it from the HTML?

Comment (guyy13) [reply student/views.py]: @guy9050 
I want to display different links to the user depending on whether they are a recruiter or a student, so I'm using an 'if' statement in the HTML body. I couldn't find a way to call the is_student() function from within the HTML so this is how I did it. If there's a better solution I'm all up for it.

File: student/views.py
Comment (ifireball) [inline student/views.py]: Do not use class based views:
https://spookylukey.github.io/django-views-the-right-way/

It generally makes the code harder to understand while providing very little benefit.

File: student/views.py
Comment (guy9050) [reply student/views.py]: Please see this message Barak posted about class-based viewes
https://redhat-beyond.slack.com/archives/C02GG3Q59D1/p1638093979130600

File: student/views.py
Comment (guyy13) [reply student/views.py]: @ifireball 
Um, ok, I'll try to recreate it using view functions..

Comment (guyy13) [reply student/views.py]: @ifireball 
Done.

File: student/views.py
Comment (guyy13) [reply student/views.py]: Fixed it

File: student/views.py
Comment (guy9050) [inline student/views.py]: meaningful variable names? i. e `student_object`

File: student/views.py
Comment (guy9050) [inline student/views.py]: update_student_form

File: student/views.py
Comment (guy9050) [inline student/views.py]: what is this for?

File: student/views.py
Comment (guyy13) [reply student/views.py]: I'll change it

File: student/views.py
Comment (guyy13) [reply student/views.py]: Same thing I explained to you on 
https://github.com/redhat-beyond/SJMaster/pull/60#discussion_r767267251

File: student/views.py
Comment (guy9050) [inline student/views.py]: I'm not sure if that's the ideal way to do it, @ifireball WDYT?

Comment (guyy13) [reply student/views.py]: @guy9050 
I found a much better way to do it,
I added a related name to the link between user and Student, and user and Recruiter, this way in the HTML I can check directly if the logged in user is a student or a recruiter without having to provide anything from the outside, I'll update it in a few minutes so you can check it out.

Comment (ifireball) [reply student/views.py]: @guy9050  Could you give me more context? What are you asking me about exactly?

@guyy13 Not sure moving logic into the HTML template is a good idea, at the very least it makes to hard to test. I would look for ways of moving it out of there. In this example, and without knowing a lot about the details, I would say that my expectation would be to have completely different page templates for student and requiters with the view selecting between them. Or perhaps even having different views and routes for each.

Comment (guyy13) [reply student/views.py]: @ifireball 
we do have different views and pages for each, the discussion here is about linking them to the navigation bar, which has to contain different links depending on who is logged in.
I just pushed the changes, please take a look.

**Comment by guyy13** [comment]
This PR became quite big so I found it right to split the commits, if you think otherwise let me know.

File: student/views.py
Comment (ifireball) [reply student/views.py]: I see this is no longer in the code, though I'm not sure how are you handling the navigation without this and I don't have the time to dive deeper right now.

When you have a set of context values that need to go into many pages in order to get embedded in shared UI items such as navigation bars, its usually a good idea to move them into a function that gets called from all the relevant views. That makes it easier to know which context items are relevant for the particular view and which are shared.

Comment (guyy13) [reply student/views.py]: @ifireball 
Ok, anyway I managed to get around this with a solution that allows you to check from within the template directly whether the logged in user is Student or a Recruiter.
I think it's simpler than creating a function like you suggested. Welcome to take a look when you have time.

Comment (ifireball) [reply student/views.py]: Ok, I see those "ifs" you have in the template. Here is the key question - how do you test that?

Comment (guyy13) [reply student/views.py]: @ifireball 
I guess I could recreate the same code in a pytest test or something?
How would I test it if I did it as you suggested (a function that contains the shared context)? I'm not sure I understand

Comment (ifireball) [reply student/views.py]: If you recreate the code in the test - you are not testing the actual code...

To make it testable - you need to move the logic out of the template - one way to do that is to write different included sub-templates for the parts that need to be different according to logic, and put the name of the part you need to include in the context. This allows you to write a test that ensures you include the right sub-template name in the context in various scenarios.

Comment (guyy13) [reply student/views.py]: @ifireball 
Ok, I'll try to do that later

**Comment by guyy13** [comment]
@ifireball @guy9050 
I added a function to the student/views.py file to add 'is_student' to the context and called it from the relevant views.
Since `is_student` is a model function of Student that also contains tests I assume it's alright.
Please let me know what you think.

**Comment by guy9050** [comment]
> @ifireball @guy9050 I added a function to the student/views.py file to add 'is_student' to the context and called it from the relevant views. Since `is_student` is a model function of Student that also contains tests I assume it's alright. Please let me know what you think.

I am not sure if that's what Barak meant, let's wait for his approval

**Comment by guyy13** [comment]
Fixed conflicts

**Comment by guyy13** [comment]
Added tests for the update form.

File: student/tests.py
Comment (guy9050) [inline student/tests.py]: This is another test where you insert invalid data, please create another test for this

Comment (guyy13) [reply student/tests.py]: fixed

File: student/views.py
Comment (ifireball) [inline student/views.py]: When you fail to find information  in a we app, you need to return HTTP 404 error code. Django has a useful `get_object_or_404` function for this:

https://docs.djangoproject.com/en/4.0/topics/http/shortcuts/

File: student/views.py
Comment (ifireball) [inline student/views.py]: Form handling code typically needs to deal with 3 scenarios:

1. When the form is initially fetched and needs to be populated with initial data from the DB
2. When form data is posted from the browser and its invalid
3. When form data is posted from the browser and its valid

Its not clear how this code deals with those scenarios, especially with the distinction between 1 and 2/3. It seems there is some assumption that the form would be invalid if its initialized from DB data only. This might be causing an unnecessary DB save.

File: student/views.py
Comment (ifireball) [inline student/views.py]: Try to be consistent and either user `render()` and `redirect()` or `TemplateResponse` and `HttpResponseRedirect`.

File: student/tests.py
Comment (ifireball) [inline student/tests.py]: This is not really testing the user flow - only the form object is being tested and that does not really contain much logic in it.

The right way to test is to consider the various flows through the form view:

1.   When the form is initially fetched, populated with initial data from the DB and displayed.
2.  When form data is posted from the browser and its invalid
3.  When form data is posted from the browser and its valid

And write a test for each that used the `client` fixture to emulate the browser activity.

**Review by ifireball — CHANGES_REQUESTED** [review]
I placed some comments. Most have to do with code style and can be fixed later, but there is a testing gap that must be fixed ASAP.

File: student/views.py
Comment (guyy13) [reply student/views.py]: I am not sure I understand what you mean here,
about 1, if I managed to get the student object from the DB, why would the form not manage to load?
About 2, the form checks the data automatically and won't let you submit it if the inserted data is invalid.
I also don't understand why the form would be invalid if it's initialized from DB data only..

Comment (ifireball) [reply student/views.py]: > if I managed to get the student object from the DB, why would the form not manage to load?

What is testing that your code is actually doing that?

>  the form checks the data automatically and won't let you submit it if the inserted data is invalid.

What checks that the view is actually checking for validity and responding properly tro the user instead of, say, throwing a nasty exception on an invalid save attempt?

What you wrote in your last comment are assumptions based on your understanding of how your code and Django works. Tests should not be making such assumptions. They should test that you get the exact behavior you need given controlled inputs without making too many assumptions on the internal implementation. Is there any guarantee that the implementation even uses proper Django Form and View objects? It might change over time. The only thing that can guarantee it works as needed is a good test.

Comment (guyy13) [reply student/views.py]: I think you mean that I add these tests in the tests file, but this is the views file so I don't understand if anything else is needed here or not.

File: student/tests.py
Comment (guyy13) [reply student/tests.py]: Okay, so let's say I'm now using the client fixture and I logged in as a student, then I make a POST request to the form, and I want to verify that the fields of the form fill out properly according to the data of the logged in user. How can I check the data that is inside the form right now without submitting it? I tried to look for it online but no luck so far.

Comment (guy9050) [reply student/tests.py]: @guyy13 you can assert the status code that is returned from the POST request

Comment (ifireball) [reply student/tests.py]: You check the form object in the context. Generally, since there is not real logic in the object - just the definition of its relationship with the model object, its may be enough to test its type to make sure the right one s displayed, but you can also check the the form object is associated with the right model object.

In the case of validation testing, you can inspect the `errors` property on the form object or call the `is_valid()` method.

See the reference for Form objects:
https://docs.djangoproject.com/en/4.0/ref/forms/api/#django.forms.Form

File: student/views.py
Comment (ifireball) [reply student/views.py]: Ok with regard to the code clarity - for example its hard to see the difference between the behavior on POST and on GET because there is explicit IF for that

File: student/tests.py
Comment (guyy13) [reply student/tests.py]: @ifireball 
I managed to get the form from the context of the page and test the type of it as you suggested, but I don't understand how to check that the fields in the form are filled out correctly with the request's user details, can you please clarify?
Thanks.
Here's my code currently:
image

Comment (guyy13) [reply student/tests.py]: Ok nvm, I think I managed to do it using `form.initial`

File: student/views.py
Comment (guyy13) [reply student/views.py]: Fixed.

File: student/tests.py
Comment (guyy13) [reply student/tests.py]: I think I managed to implement the tests, please take a look.

File: student/tests.py
Comment (ifireball) [inline student/tests.py:158]: Missing test for the redirection target - there is an HTTP header that contains it in the response.

**Review by ifireball — APPROVED** [review]
Added a small comment, otherwise its looks good enough.

File: student/tests.py
Comment (guyy13) [reply student/tests.py:158]: Done!

**Comment by guy9050** [comment]
looks much better now!

**Pull request merged**
